### ISSUE TITLE
About Tensorboard

### ISSUE THREAD

**Issue by Nara0731** [openai/baselines#596]
Hello,
I am not sure about how to set a Tensorboard?
I have set the environment variable, 
I use the PyCharm, so I donot know how to modify the --log-dir ?

**Comment by Nara0731** [comment]
It shows "Logging to /tmp/openai-2018*". However, I canno find the directory "/tmp".

**Comment by Nara0731** [comment]
I have found the directory “/tmp". However, I only find
0.0.monitor.csv
log.txt
progress.csv
Where should I find the file about Tensorboard？

**Comment by pzhokhov** [comment]
Hi @Nara0731 ! We have recently added this section to the README: https://github.com/openai/baselines/blob/master/README.md#using-baselines-with-tensorboard
basically, you need to set env variables:
OPENAI_LOGDIR to where you want the tensorboard files to be saved, and 
OPENAI_LOG_FORMAT to 'stdout,tensorboard' (if you only need output to command line and tensorboard). 
The tensorboard data should show up in OPENAI_LOGDIR (subfolder tb). 
You can launch tensorboard via `tensorboard --logdir=$OPENAI_LOGDIR`
From the fact logs are saved to /tmp/openai-2018* location, I suspect that neither of the environment variables is actually set (at least from python's interpreter perspective). Could you run 
```
import os; print(os.environ)
``` 
in python and paste here the output? If OPENAI_LOGDIR and OPENAI_LOG_FORMAT are not there, you can set them directly from python:
```
os.environ['OPENAI_LOGDIR'] = ...
os.environ['OPENAI_LOG_FORMAT'] = 'stdout,tensorboard'
```
(that has to happen before you start training)

**Comment by Nara0731** [comment]
Yeah， I have set 
export OPENAI_LOG_FORMAT='stdout,log,csv,tensorboard' # formats are comma-separated, but for tensorboard you only really need the last one
export OPENAI_LOGDIR=/tmp

Unfortunately, I did not find any relevant file about tensorboard in the "/tmp"

**Comment by pzhokhov** [comment]
hm... let's solve it one step at a time. Could you run `import os; print(os.environ)` in python?

**Comment by pzhokhov** [comment]
Thanks! Yeah, so basically one way or another the OPENAI_LOGDIR and OPENAI_LOG_FORMAT do not make it to the python process environment variables. The fix is really easy - add
```
import os
os.environ['OPENAI_LOGDIR']='/tmp'
os.environ['OPENAI_LOG_FORMAT']='stdout,tensorboard'
```
to the very top of your python script; and try running it again. Ideally, tensorboard checkpoints should show up in `/tmp/tb` folder. Please let me know if that does not work for you,

**Comment by Nara0731** [comment]
I  cannot find the "/tmp/tb", I only find "tmp"

**Comment by pzhokhov** [comment]
okay; could you post here your python code please? Thanks!

**Comment by smalltingting** [comment]
Hi, I have the same problem as you.
I solved the problem like this.
Just modify the code in run.py.
`    if MPI is None or MPI.COMM_WORLD.Get_rank() == 0:
        rank = 0
        logger.configure(dir='./log',format_strs=['stdout','log','csv','tensorboard'])`

**Comment by smalltingting** [comment]
> Hello,
> I am not sure about how to set a Tensorboard?
> I have set the environment variable,
> I use the PyCharm, so I donot know how to modify the --log-dir ?



> 您好，
> 我不确定如何设置Tensorboard？
> 我已经设置了环境变量，
> 我使用PyCharm，所以我不知道如何修改--log-dir？
Hi, I have the same problem as you.
I solved the problem like this.
Just modify the code for 209th in run.py.
`if MPI is None or MPI.COMM_WORLD.Get_rank() == 0: 
      rank = 0 
      logger.configure(dir='./log',format_strs=['stdout','log','csv','tensorboard']`

**Comment by Nara0731** [comment]
Really? I will try it.

**Comment by srivatsankrishnan** [comment]
Hi @pzhokhov @smalltingting 
I configured the logger setting as you have mentioned. I see it created a directory called "tb". However, it is empty. Any idea what is going on?

I am using deepq example but I think it shouldn't matter. This is how I configure it in my code:
```
def main():
    logger.configure(dir='.log', format_strs=['stdout', 'log', 'csv', 'tensorboard'])
```

**Comment by pzhokhov** [comment]
@srivatsankrishnan does logger print anything on the screen / in the log file? Logger only saves data when a logger.dumpkvs() (or logger.dump_tabular()) is called, which by default happens fairly rarely in deepq. Could you try with `--print_freq=1` option?

**Comment by srivatsankrishnan** [comment]
Hi @pzhokhov,
The only thing the logger prints in the screen is this message:
"Logging to .log"

It creates the following folder structure in .log:
/logs
     |------tb
     |------log
     |------progress

The tb folder is empty. The progress.csv is also empty. The "log" ( the file that gets created inside the directory) file basically has the same message that was printed in the console ("Logging to .log"). I tried changing the --print_freq=1 but the results are the same.

I tried to hack the code where I create my model (models.py) to explicitly export my graph to visualize in TensorBoard. 
This is what i use:
```
tf_writer = tf.summary.FileWriter(LOGDIR)
tf_writer.add_graph(tf.get_default_session().graph)
```
But the graph is too complex and can't trace to my input and output nodes ( Honestly trying to make sense of it and not given on that yet). I assume the functionality that you guys enable with logger for tensorboard will be more structured or methodical to visualize it in tensorboard.

**Comment by pzhokhov** [comment]
Hi @srivatsankrishnan ! Sorry about the lag. If all the progress.csv is empty, tb/ subfolder is empty and nothing interesting is printed on the screen, it means that 
1) the training did not progress to the point where it would save anything (call logger.dump_tabular()) . 
or 2) something bad happened to the logger module

Could you try running a simple test with deepq, for instance:
```
export OPENAI_LOG_FORMAT=stdout,csv,tensorboard
export OPENAI_LOGDIR=.log
python -m baselines.run --alg=deepq --env=CartPole-v0 --print_freq=1 --num_timesteps=1e5
```
If everything works correctly, this should generate a long output that looks like:
```
-----------------------------------
| % time spent exploring      | 67    |
| episodes                             | 1242  |
| mean 100 episode reward | 28.9  |
| steps                                   | 32989 |
-----------------------------------
```
and files `progress.csv`, `0.0.monitor.csv`, `log.txt`, and subfolder `tb` in .log. 
If that works, but your case still does not, it probably means that logger / logger configuration are messed up. If the test above does not work, then something in your python environment is not quite right; and in that case, I'd recommend installing baselines in a clean virtualenv, and trying again.

**Comment by srivatsankrishnan** [comment]
Hi @pzhokhov!  No worries. This one works and I see logs and event file getting generated. When I open tensorboard, it only has the scalars such as (% time spent exploring, episodes, rewards etc). I don't see a graph in tensorboard.I was interested in seeing the graph for the neural net model to determine input and output nodes.  I just hacked the code where I define the model to capture the graph. So in a way, I was able to get what I wanted.

As you put it, In my case, I just put 100 steps for my environment and --print_freq=1 to quickly capture the graph. Maybe it didn't get to a point where logger.dump_tabular() wasn't getting called. 

On a different note, Is there a plan to support saving the model in native tensorflow format along with graph (.pb)? The reason is that there are lots of interesting tools in tensorflow and they basically require the model in one of these formats.

**Comment by srivatsankrishnan** [comment]
Hi @pzhokhov,
Thanks for your reply. There are lots of tools in tensorflow to fine-tune inference performance:
(https://github.com/tensorflow/tensorflow/tree/master/tensorflow/python/tools)(https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/graph_transforms/README.md)
The basic requirement is to use these tools is to have models saved in native tensorflow format (checkpoints, .pb etc). I am particularly interested in using these tools and was able to hack the code to save it in the native tensorflow format. I am currently facing some tensorflow related issue but will soon be able to test it out once I resolve those.

I have one more useful functionality in mind but its orthogonal to this discussion. Maybe I will open a new issue for it to avoid mixing it up with this.

**Comment by RyanRizzo96** [comment]
@pzhokhov, when I try what you suggest everything works fine. I test the following script successfully.

```
export OPENAI_LOG_FORMAT=stdout,csv,tensorboard
export OPENAI_LOGDIR=.log
python -m baselines.run --alg=deepq --env=CartPole-v0 --print_freq=1 --num_timesteps=1e5
```

I am confused however because if I replace the algorithm and environment to:

`python3 -m baselines.run --alg=her env=FetchReach-v1 --print_freq=1 --num_timesteps=1e5`

I get an error:

> 
>   File "/Users/ryanr//code/baseline_code/baselines/baselines/her/experiment/config.py", line 192, in configure_dims
>     'o': obs['observation'].shape[0],
> IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices


Any idea why? Cheers!

**Comment by ryanmaxwell96** [comment]
Was there any resolution to this issue? I've tried the same suggestions that have been listed so far (os.environ['OPENAI_LOGDIR'] = ... and os.environ['OPENAI_LOG_FORMAT] = 'stdout,tensorboard') and I can get those to be listed on print(os.environ), but I am not getting any file outputs. Any ideas?

**Comment by KomputerMaster64** [comment]
I just wanted to know how to move the logs from `/tmp` directory to directory of choice, as I have to manually save the `/tmp/openai-2022.....` files to get the checkpoints for training.

PS I am using multiple gpus for training, hope that suggested methods works for multi gpu training

**Comment by Nara0731** [comment]
你好，我已收到你的来信。若有重要事情，请短信告知！
